src/fix_efield.cpp

```diff
@@ -449,3 +452,70 @@ double FixEfield::compute_vector(int n)
   }
   return fsum_all[n+1];
 }
+
+/* ----------------------------------------------------------------------
+   get E
+------------------------------------------------------------------------- */
+
+double* FixEfield::get_energy()
+{
+  double **x = atom->x;
+  int *mask = atom->mask;
+  imageint *image = atom->image;
+
+  int nlocal = atom->nlocal;
+
+  // reallocate energy array if necessary
+
+  if (atom->nmax > maxatom_energy) {
+    maxatom_energy = atom->nmax;
+    memory->destroy(energy);
+    memory->create(energy,maxatom_energy,"efield:energy");
+  }
+  memset(&energy[0],0.0,maxatom_energy*sizeof(double));
+
+  // update region if necessary
+
+  Region *region = NULL;
+  if (iregion >= 0) {
+    region = domain->regions[iregion];
+    region->prematch();
+  }
+
+  int warn_flag_local = 0;
+
+  // constant efield
+
+  if (varflag == CONSTANT) {
+    double unwrap[3];
+
+    // charge interactions
+    // force = qE, potential energy = F dot x in unwrapped coords
+
+    if (qflag) {
+      for (int i = 0; i < nlocal; i++) {
+        if (mask[i] & groupbit) {
+          if (region && !region->match(x[i][0],x[i][1],x[i][2])) continue;
+          const double fx = ex;
+          const double fy = ey;
+          const double fz = ez;
+
+          domain->unmap(x[i],image[i],unwrap);
+          energy[i] -= fx*unwrap[0] + fy*unwrap[1] + fz*unwrap[2];
+          if (fabs(fx*(x[i][0]-unwrap[0])) + fabs(fy*(x[i][1]-unwrap[1])) +
+              fabs(fz*(x[i][2]-unwrap[2])) > 0.0)
+            warn_flag_local = 1;
+        }
+      }
+    }
+  } else {
+    error->all(FLERR,"Cannot yet use fix qeq/reaxff with variable efield");
```

Comment:
@akohlmey @sjplimp is is OK to have this QEq specific code in `fix_efield.cpp`? This is the only sticking point I can think of in this PR... I could move this into `fix_qeq` but it wouldn't be very convenient.

Reply:
@stanmoore1 I am not so much concerned about the fact that this function is called (there should be some more detailed comments about what it does and when it is called by what), but I am more worried about using unwrapped coordinates. 
Having an electric field across PBC is a bit of a problem. An atom that will pass through the PBC will see a jump in the potential, but if you use unwrapped coordinates the atoms will disperse and atoms that are on one level neighbors are not on the other. With planewave DFT codes the "solution" usually is to assume a sawtooth shaped potential and require a vacuum gap in where the potential "snaps" back and no atoms or electrons may go. But even assuming the approach used here is documented and its inconsistency accepted, I don't understand why there then have to be warnings for non-zero image flags (and why that is tested for in such a complex way)? For PBC this is practically a given to have non-zero image flags and for non-PBC it is not a problem (we already warn for that and zero them), and unlike planewave DFT we don't need to force PBC for a slab system with ReaxFF because it doesn't use kspace. So I am wondering, am I misssing something?

Reply:
This electric field code should not be used with ReaxFF when there is a periodic boundary in the field direction, hence the warning. Are you saying you zero image flags whenever a periodic BC is converted into a non-periodic BC?

Reply:
If LAMMPS always zeros out image flags when we can just convert the check into a check for PBC in the field direction...

Reply:
I think that is what happens when you use read_data and change_box. This was changed not so long ago, though. But if you want to disallow the use of the field in the way you describe, having a warning here is not a good solution. Starting with the fact that it is difficult to interpret the warning and that it may accumulate a lot of output. How about adding an "extract" function and being able to query the fix for the field vector components and whether it is constant or not and then make a check for PBC (via `domain->periodic` or similar) during Pair::init_style() in case the field fix is active? Then the calculation could simply terminate with a more descriptive error message ("must not have periodicity in ? direction with field vector <x,y,z>").

Reply:
Yes, I think this is a better solution.

Reply:
@stanmoore1 do you want me to try and implement this or would prefer to do this yourself?

Reply:
If you have cycles to work on this, that would be great, otherwise I can do it. I think adding an "extract" method to `fix efield` as you suggested and then moving all the field code into fix `qeq/reaxff` is the cleanest. It shouldn't require any unwrapped coordinates if we can confirm that image flags cannot be non-zero if a PBC is change to a non-PBC, then add an error check for PBC in the same direction as the field.

Reply:
I am not subject to the kind of constraints that people in national labs are. That is why I asked in the first place. In my case this just means something else has to wait a little longer. My list of things that I could work on is always very long... Just look at the "feature request" issues that I created recently. and that is only the tip of the iceberg. :wink:

Reply:
Thus I will have a look at this. Perhaps over the weekend. I don't want this PR to linger around much longer because of some of the more sweeping changes planned (see e.g. issue #2946) that will then create extra work merging pull requests, but already merged changes will be processed "for free". Converting the PR back to a draft for the time being.

Reply:
Thanks!